scholarly/_navigator.py

```diff
@@ -0,0 +1,195 @@
+        super(Navigator, self).__init__()
+        logging.basicConfig(filename='scholar.log', level=logging.INFO)
+        self.logger = logging.getLogger('scholarly')
+
+
+    def _get_page(self, pagerequest:str):
+        """Return the data for a page on scholar.google.com"""
+        self.logger.info("Getting %s", pagerequest)
+        resp = None
+        while True:
+            # If Tor is running we use the proxy
+            # Did not use with for shorter indented lines -V
+            session = requests.Session()
+            if self._tor_works():
+                # I *really* dislike that we do two requests every time
+                # checking if tor is working after connection is established
+                # should only happen when errors happen.
+
+                # Tor uses the 9050 port as the default socks port
+                # on windows 9150 for socks and 9151 for control
+                session.proxies = {'http':  _TOR_SOCK,
+                                   'https': _TOR_SOCK}
+            else:
+                session.proxies = _PROXIES
+
+            try:
+                _HEADERS['User-Agent'] = UserAgent().random
+
+                resp = session.get(pagerequest,
+                                   headers=_HEADERS,
+                                   cookies=_COOKIES,
+                                   timeout=_TIMEOUT)
+                
+                if resp.status_code == 200:
+                    if self._has_captcha(resp.text):
+                        raise Exception("Got a CAPTCHA. Retrying.")
+                    else:
+                        break
+                else:
+                    self.logger.info(f"""Got a response code {resp.status_code}. 
+                                    Retrying...""")
+                    raise
+
+            except Exception as e:
+                self.logger.info(f"Exception {e} while fetching page. Retrying.")
+                # Check if Tor is running and refresh it
+                self.logger.info("Refreshing Tor ID...")
+                self._refresh_tor_id()
+                session.close()
+        session.close()
+        return resp.text
+
+
+    def _tor_works(self) -> bool:
+        """ Checks if Tor is working"""
+        with requests.Session() as session:
+            session.proxies = {
+                'http': _TOR_SOCK,
+                'https': _TOR_SOCK
+            }
+            try:
+                #Changed to twitter so we dont ping google twice every time
+                resp = session.get("http://www.twitter.com")
+                self.logger.info("TOR Works!")
+                return resp.status_code == 200
+            except Exception as e:
+                self.logger.info("NO TOR :(")
+                return False
+
+
+    def _refresh_tor_id(self) -> bool:
+        try:
+            with Controller.from_port(port=_TOR_CONTROL) as controller:
+                controller.authenticate(password="")
```

Comment:
When using the `./setup_tor.sh` we use the password `scholarly_password`. We should either make this configurable for both the `./setup_tor.sh` and this part of the code, or use `scholarly_password` for both.